Add CachingLeafSlicesSupplier to compute the LeafSlices for concurrent segment search

sohami opened the pull request:
### Description
Add a constructor which takes in the computed slices from extensions and uses that for running the search concurrently on provided executor. This is based on the discussion on the issue https://github.com/apache/lucene/issues/12347

sohami commented:
@javanna Can you help to look into this PR ? Would be great if we can merge it in before release branch cut off.

andrross commented on lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java:
```diff
@@ -1984,14 +1985,16 @@ protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
               };
         }
-      } else {
+      } else if (randomBoolean()) {
```
I think `random.nextBoolean()` would be preferred here to match the existing pattern.

andrross replied:
Also, for completeness, should the same new constructors be added to [AssertingIndexSearcher](https://github.com/apache/lucene/blob/main/lucene/test-framework/src/java/org/apache/lucene/tests/search/AssertingIndexSearcher.java)?

javanna commented:
hi @sohami thanks for opening this PR! I have been considering a different approach, in the attempt of not adding additional constructor to IndexSearcher, as it already has quite some.

I am thinking that pre-computing the slices in the constructor may not be so important. How about removing that and computing the slices each time they are retrieved? There's only three methods where that is used between searchAfter and search. If needed, subclasses can still pre-compute and cache the slices if that's considered important. This way, subclasses are free to override how slices are computed, and the change is a little less invasive.

sohami commented:
> hi @sohami thanks for opening this PR! I have been considering a different approach, in the attempt of not adding additional constructors to IndexSearcher, as it already has quite some.
> 
> I am thinking that pre-computing the slices in the constructor may not be so important. How about removing that and computing the slices each time they are retrieved? There's only three methods where that is used between searchAfter and search. If needed, subclasses can still pre-compute and cache the slices if that's considered important. This way, subclasses are free to override how slices are computed, and the change is a little less invasive.

@javanna Thanks for the suggestion. I would think the base implementation will want to avoid computing it on every usage specially for the immutable indices where [IndexSearcher is suggested to be reused](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L54). If it is not cached then for each search it will compute the slices which seems redundant and also deviating from current baseline. If we want to cache the computed slices in base class but remove it from getting initialized from constructor, then essentially we are making `leafSlices` variable non-final and relying on users of the `leafSlice` variable across the methods to use `getSlices()` atleast once which doesn't seem clean as well.

I think if the concern is to avoid adding yet another constructor then other way could be to make the `leafSlice` variable `protected` and `non-final` and initialize it in constructor of base class. The subclasses can then choose to override it as needed with the understanding that it should not be changed in between the execution (not clean but probably a middle ground and similar to how `similarity` implementation is provided to `IndexSearcher`)

sohami commented:
@javanna I am also thinking with regards to too many constructor can we see if deprecating the ones with IndexReader as inputs and keeping the ones with IndexReaderContext over the time will make sense. For the new one as well we can avoid adding one with IndexReader if we go down that path vs non final variable.

jpountz commented:
I'm not entirely happy with this new constructor either, because now there are two ways to customize slicing: either with this new constructor or by overriding the `slices()` method. I'd like to have a single one, and the approach to override a method to compute slices given a list of segments looks more user-friendly to me than having to compute slices up-front?

For reference, this [20-years-old comment](https://github.com/apache/lucene/commit/360d91dde75244b3b73793ac3a2e6c8737c45f9f) on `IndexSearcher` about making sure to reuse `IndexSearcher` instances is outdated, all IndexSearcher constructors are cheap compared to the work that running any query would need to do. It's actually required to create a new `IndexSearcher` for every search if you wish to leverage `IndexSearcher`'s timeout support.

If there is reluctance to computing slices on every call to `#search`, could we compute slices lazily and cache them in `getSlices()`? This way, it wouldn't be needed to increase the visibility of `slices` to `protected`?

andrross commented:
+1 to not having two ways to customize slicing.

> the approach to override a method to compute slices given a list of segments looks more user-friendly to me than having to compute slices up-front?

@jpountz I agree. I think the basic problem here is that the protected method is called from the constructor of the parent class, which really limits the flexibility and can lead to unexpected behavior. The simplest approach is to just compute the slices on every call to `#search`. If we want to be more defensive about possible performance regressions, then lazy computation + caching like you said is the way to go (i.e. memoization). Is there any utility in Lucene like this [memoizing supplier](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Suppliers.java#L157-L201) in Guava? (I couldn't find one but I'm not super familiar with the code base). Hand rolling the memoization logic inside IndexSearch is an option too, we'll just need to take some care to ensure that it is thread safe.

sohami commented:
> +1 to not having two ways to customize slicing.

>> the approach to override a method to compute slices given a list of segments looks more user-friendly to me than having to compute slices up-front?

Agreed to this however the `slices` method doesn't provide got way to overwrite the computation for all the cases as it is called from base class constructor.

> For reference, this [20-years-old comment](https://github.com/apache/lucene/commit/360d91dde75244b3b73793ac3a2e6c8737c45f9f) on IndexSearcher about making sure to reuse IndexSearcher instances is outdated, all IndexSearcher constructors are cheap compared to the work that running any query would need to do. It's actually required to create a new IndexSearcher for every search if you wish to leverage IndexSearcher's timeout support.

Thanks for sharing the context

> If there is reluctance to computing slices on every call to #search, could we compute slices lazily and cache them in getSlices()? This way, it wouldn't be needed to increase the visibility of slices to protected?

With `getSlices()` based approach and using member variable `leafSlices` to cache the computed value, my only concern is that it will be hard to enforce any usage to fetch the slices within `IndexSearcher` to use `getSlices` always (and not the cached member directly). With that in mind I guess its cleaner to compute `slices` always and remove the member `leafSlices`. Thoughts ?

jpountz commented:
> Is there any utility in Lucene like this [memoizing supplier](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Suppliers.java#L157-L201) in Guava? (I couldn't find one but I'm not super familiar with the code base). Hand rolling the memoization logic inside IndexSearch is an option too, we'll just need to take some care to ensure that it is thread safe.

No, Lucene doesn't have such a utility. +1 to hand rolling the memoization logic. I don't think it needs to be fancy in terms of concurrency, IndexSearcher is expected to spend much more time scoring hits than computing slices, so making it synchronized should be good enough?

> With that in mind I guess its cleaner to compute slices always and remove the member leafSlices. Thoughts ?

I am happy either way. Caching feels like it has less potential for being perceived as a regression by some users. One approach could be to do the caching approach now which is a bit safer, and then move to the approach that you're suggesting on the `main` branch only so that it goes in the next major where changes in runtime behavior would be less surprising. I don't have strong feelings about this.

> Also wondering what is the disadvantage in making leafSlices, protected and non final.

I don't like this approach because it exposes internals of IndexSearcher than I'd like users to not touch or have to know about. For instance if we made it protected and non-final, and later made a change to compute slices lazily instead of eagerly in the constructor, this could be perceived as a breaking change as there would be cases when leafSlices would now be `null` when it wasn't before.

sohami committed:
Add a thread safe CachingLeafSlicesSupplier to compute and cache the LeafSlices used with concurrent segment
search. It uses the protected method `slices` by default to compute the slices which can be
overriden by the sub classes of IndexSearcher

sohami commented:
@jpountz @andrross As discussed, I have updated the PR with a thread safe caching LeafSlices supplier implementation. The supplier uses the `slices` method to compute the leaf slices

jpountz commented on lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:
```diff
@@ -116,8 +117,11 @@ public class IndexSearcher {
+   * used with executor - LeafSlice supplier where each slice holds a set of leafs executed within
+   * one thread
+   */
+  private final CachingLeafSlicesSupplier leafSlicesSupplier;
```
Can you add a comment that we're caching it instead of creating it eagerly to avoid calling a protected method from a constructor, which is a bad practice?

jpountz commented on lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:
```diff
@@ -1014,4 +1021,48 @@ private static SliceExecutor getSliceExecutionControlPlane(Executor executor) {
+    public LeafSlice[] get() {
+      if (executor == null) {
+        return null;
+      }
```
It would make more sense to have to have the `executor == null` check in `IndexSearcher#getSlices` than here so that this class focuses solely on caching logic?

sohami replied:
I did it this way because `IndexSearcher` can access `leafSlicesSupplier` without using `getSlices`. So everywhere it will need to perform explicit check for executor being `non-null` too for such usages. With current mechanism non null check for executor will not be needed for any future usages as well. This is also keeping the current behavior to return `null` slices when `executor` is `null` as well.

sohami replied:
@jpountz I have moved it to `getSlices` method and using that everywhere as you suggested earlier. If `executor` is `null` then initializing the supplier as `null` in that case. Please let me know if this looks good.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:
```diff
@@ -1014,4 +1021,48 @@ private static SliceExecutor getSliceExecutionControlPlane(Executor executor) {
+   * passed to the caller of this supplier on each invocation. If the provider returns null then
+   * {@link NullPointerException} will be thrown to the caller.
+   */
+  private static class CachingLeafSlicesSupplier implements Supplier<LeafSlice[]> {
```
Maybe also add a comment that this class is implementing the (subtle) double-checked locking idiom and link to something like https://shipilev.net/blog/2014/safe-public-construction/ that describes the traps of it?

sohami commented:
@jpountz Gentle reminder for the review

jpountz reviewed (approved):
LGTM. Can you add a CHANGES entry under 9.8?

sohami commented:
> LGTM. Can you add a CHANGES entry under 9.8?

Added. I have also rebased the changes. Thanks @jpountz for your time and guidance.
